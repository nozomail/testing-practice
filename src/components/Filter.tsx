import React from "react";
import { FiltersProps } from "../types";

type Props = {
  filters: FiltersProps;
  setfilters: React.Dispatch<React.SetStateAction<FiltersProps>>;
};

function Filter({ filters, setfilters }: Props) {
  function handleChange(type: "alcoholic" | "nonAlcoholic") {
    switch (type) {
      case "alcoholic":
        return setfilters({
          ...filters,
          isAlcoholic: !filters.isAlcoholic,
        });

      case "nonAlcoholic":
        return setfilters({
          ...filters,
          isNonAlcoholic: !filters.isNonAlcoholic,
        });
      default:
        return;
    }
  }

  return (
    <div className="Filter">
      <input id="alcoholic" type="checkbox" checked={filters.isAlcoholic} onChange={() => handleChange("alcoholic")} />
      <label htmlFor="alcoholic">Alcoholic</label>

      <input id="nonAlcoholic" type="checkbox" checked={filters.isNonAlcoholic} onChange={() => handleChange("nonAlcoholic")} />
      <label htmlFor="nonAlcoholic">Non Alcoholic</label>
    </div>
  );
}

export default Filter;
