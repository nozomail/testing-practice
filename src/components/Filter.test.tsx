import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Filter from "./Filter";

describe("Filter", () => {
  test("renders all filter items checked", () => {
    const setfilters = jest.fn();
    const filters = {
      isAlcoholic: true,
      isNonAlcoholic: true,
    };

    render(<Filter filters={filters} setfilters={setfilters} />);

    const filter1 = screen.getByRole("checkbox", { name: "Alcoholic" });
    const filter2 = screen.getByRole("checkbox", { name: "Non Alcoholic" });

    expect(filter1).toBeInTheDocument();
    expect(filter2).toBeInTheDocument();
    expect(filter1).toBeChecked();
    expect(filter2).toBeChecked();
  });

  test("calls setfilter with proper args when Alcoholic filter is clicked", () => {
    const setfilters = jest.fn();
    const filters = {
      isAlcoholic: true,
      isNonAlcoholic: true,
    };

    render(<Filter filters={filters} setfilters={setfilters} />);

    const filter1 = screen.getByRole("checkbox", { name: "Alcoholic" });

    userEvent.click(filter1);

    expect(setfilters).toBeCalledWith({
      isAlcoholic: false,
      isNonAlcoholic: true,
    });
    expect(setfilters).toHaveBeenCalledTimes(1);
  });

  test("calls setfilter with proper args when Non Alcoholic filter is clicked", () => {
    const setfilters = jest.fn();
    const filters = {
      isAlcoholic: true,
      isNonAlcoholic: true,
    };
    render(<Filter filters={filters} setfilters={setfilters} />);
    const filter1 = screen.getByRole("checkbox", { name: "Non Alcoholic" });

    userEvent.click(filter1);

    expect(setfilters).toBeCalledWith({
      isAlcoholic: true,
      isNonAlcoholic: false,
    });
    expect(setfilters).toHaveBeenCalledTimes(1);
  });
});
