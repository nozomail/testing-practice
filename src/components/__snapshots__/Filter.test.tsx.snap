// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter mathes snapshot 1`] = `
<DocumentFragment>
  <div
    class="Filter"
  >
    <input
      checked=""
      id="alcoholic"
      type="checkbox"
    />
    <label
      for="alcoholic"
    >
      Alcoholic
    </label>
    <input
      checked=""
      id="nonAlcoholic"
      type="checkbox"
    />
    <label
      for="nonAlcoholic"
    >
      Non Alcoholic
    </label>
  </div>
</DocumentFragment>
`;
